### ISSUE TITLE
improved `calc` construct

### ISSUE THREAD

**Issue by leodemoura** [leanprover/lean#268]
Soonho and I discussed the following improvements for `calc`.
Given,

```
calc a_1 = a_2 : H_1
    ...  = a_3 : H_2
    ...
```

here is the list of improvements:
1- `H_i`'s  must be elaborated after `a_i`'s have been elaborated
2- `symm` is automatically inserted if needed
3- `!` is automatically inserted if needed
4- `{...}` is automatically inserted if needed
5- By default, `{...}` is implemented by using higher-order unification. This is expensive, and an overkill in many instances. Our goal is to try first the quick and simple substitution.

With these modifications, the following proof

```
definition nt_compose (η : G ⟹ H) (θ : F ⟹ G) : F ⟹ H :=
natural_transformation.mk
  (take a, η a ∘ θ a)
  (take a b f, calc
    H f ∘ (η a ∘ θ a) = (H f ∘ η a) ∘ θ a : !assoc
                  ... = (η b ∘ G f) ∘ θ a : {naturality η f}
                  ... = η b ∘ (G f ∘ θ a) : !assoc⁻¹
                  ... = η b ∘ (θ b ∘ F f) : {naturality θ f} 
                  ... = (η b ∘ θ b) ∘ F f : !assoc)
```

can be written as

```
definition nt_compose (η : G ⟹ H) (θ : F ⟹ G) : F ⟹ H :=
natural_transformation.mk
  (take a, η a ∘ θ a)
  (take a b f, calc
    H f ∘ (η a ∘ θ a) = (H f ∘ η a) ∘ θ a : assoc
                  ... = (η b ∘ G f) ∘ θ a : naturality η f
                  ... = η b ∘ (G f ∘ θ a) : assoc
                  ... = η b ∘ (θ b ∘ F f) : naturality θ f
                  ... = (η b ∘ θ b) ∘ F f : assoc)
```

**Comment by fpvandoorn** [comment]
Sounds good! I'm a little bit about automatically inserting `{...}` because currently it's so expensive. Hopefully this is not a problem if first the simple substitution is tried.

**Comment by leodemoura** [comment]
> I'm a little bit about automatically inserting {...} because currently it's so expensive. 

This is a good point. Perhaps, we should only try `{...}` if it is an easy case (e.g., `higher-order pattern matching`). This is case in most instances of `{...}` in the standard library.

**Comment by avigad** [comment]
Will this work by inserting backtracking points? My concern is that if
there is a typo somewhere in a long proof, the elaborator will try all the
useless combinations of attempts in a calc section, and then give a
nonsense error message.

Also, sometimes the justifications in a calculation proof can be very long
(think of calculations with sums and limits and integrals, etc., etc.),
like here:

  https://github.com/avigad/isabelle/blob/master/Analysis/Levy.thy#L403-L492

Could calculations like these potentially become problematic with the new
features?

On Mon, Oct 27, 2014 at 3:56 PM, Leonardo de Moura <notifications@github.com

**Comment by leodemoura** [comment]
@avigad, I think the performance may improve.
Given,

```
calc a_1 = a_2 : H_1
    ...  = a_3 : H_2
    ...
```

The idea is to elaborate `a_i`s before `H_i`s. When all `a_i`s have been elaborated, the calc construct will try to elaborate each `H_i`. The different interpretations will be tried in sequence.
There is no backtracking point. 

That being said, this new semantics may break some files. For example, any file where the elaboration of `a_i`s is influenced/restricted by the `H_i`s may potentially break.
I don't think this is a big issue. Moreover, I think it is bad practice to have a statement that depends on its proof to be elaborated. That is, the meaning depends on the proof.

**Comment by avigad** [comment]
This sounds good to me.

**Comment by leodemoura** [comment]
I tried to identify places in the standard library where the proof influences/constraints what is being proved. I found only one instance at quotient/basic.lean (https://github.com/leanprover/lean/commit/407e35692b25b02ca12193b86890ce1fccdc7394)

**Comment by fpvandoorn** [comment]
This change is very nice! 

Just a note: in some cases, removing the braces `{...}` is very expensive. 
In the following definition, when I remove the indicated braces, the compilation takes ~3 seconds (with braces it is almost instantaneous)
(at least in my personal repository).

```
  protected definition compose (G : functor D E) (F : functor C D) : functor C E :=
  functor.mk
    (λx, G (F x))
    (λ a b f, G (F f))
    (λ a, calc
      G (F (ID a)) = G id : {respect_id F a} -- <-- here
               ... = id   : respect_id G (F a))
    (λ a b c g f, calc
      G (F (g ∘ f)) = G (F g ∘ F f)     : respect_comp F g f
                ... = G (F g) ∘ G (F f) : respect_comp G (F g) (F f))
```

**Comment by leodemoura** [comment]
Could you double check it?
On my machine, it takes 0.5 secs to process the whole file with `{}`, and 0.56 secs without.
I'm not using any form of caching. Are you using the `debug` or `release` build?

Thanks,
Leo

**Comment by leodemoura** [comment]
I fetched you repository, and managed to reproduce the issue.
I will investigate the performance problem.

**Comment by leodemoura** [comment]
@fpvandoorn I have been investigating this performance issue, and I'm still trying to understand what is going on. It seems the new `calc` command is producing a sequence of constraints that exposes an inefficiency in the higher-order unifier. 

The source of the problem does not seem to be the new `calc` command, but the order the constraints are solved. I tried different variations of the line

```
calc G (F (ID a)) = G id : {respect_id F a}
```

with very disturbing behavior. The difference in performance between "fast" and "slow" is (on my home machine): 0.2 secs (fast) and 1.6 secs (slow).
If we use the line above, it is fast. The following variations are all slow

```
      calc G (F (ID a)) = G id : respect_id F a                                  
      calc G (F (ID a)) = G id : eq.subst (respect_id F a) rfl                   
      calc G (F (ID a)) = G id : eq.subst (respect_id F a) (eq.refl (G (F (ID a))))
```

In the first variation, we are relying on the new calc command to introduce the `subst`, but this is not the case for the 2nd and 3rd variations. The 3rd is disturbing since it is "almost" identical to what `{repect_id F a}` expands too. The main difference is that when we use `{...}`, we don't create two copies of `(G (F (ID a)))`. 

Then, I tried the following variation

```
let lhs := G (F (ID a)) in calc lhs = G id : eq.subst (respect_id F a) (eq.refl lhs)
```

This one is more faithful that what is generated by the `{...}` notation because it does not create two copies of `G (F (ID a))`.
Surprisingly, this one is fast. It is as fast as `{respect_id F a}`.

I also tried to put an extra delay step using `proof-qed`. 

```
calc G (F (ID a)) = G id : proof eq.subst (respect_id F a) (eq.refl (G (F (ID a)))) qed
calc G (F (ID a)) = G id : proof eq.subst (respect_id F a) rfl qed
```

Surprisingly, both of them are fast. The also show that delaying the unification constraints is not the trigger for the performance problem.
It seems the only possible explanation is the order the constraints are solved. 

Moreover, the problem is not really the new `calc` expression, but some performance problem in the higher-order unification procedure. 

Yesterday, I was working on the `structure` command, I will return to it today.
I will close this issue and open a new performance issue for the higher-order unifier.
